zerver/tests/test_scheduled_messages.py

```diff
@@ -81,6 +89,203 @@ def test_schedule_message(self) -> None:
+        logger.info.assert_has_calls(calls)
+        self.assertEqual(logger.info.call_count, 2)
+        self.assertTrue(scheduled_message.failed)
+
+    def test_too_late_to_deliver_scheduled_message(self) -> None:
+        logger = mock.Mock()
+        self.create_scheduled_message()
+        scheduled_message = self.last_scheduled_message()
+
+        too_late_to_send_message_datetime = (
+            scheduled_message.scheduled_timestamp
+            + datetime.timedelta(minutes=SCHEDULED_MESSAGE_LATE_CUTOFF_MINUTES + 1)
+        )
+        with mock.patch(
+            "zerver.actions.scheduled_messages.timezone_now",
+            return_value=too_late_to_send_message_datetime,
+        ):
+            self.verify_deliver_scheduled_message_failure(scheduled_message, logger)
+
+    def test_realm_deactivated_failed_to_deliver_scheduled_message(self) -> None:
+        logger = mock.Mock()
+        self.create_scheduled_message()
+        scheduled_message = self.last_scheduled_message()
+
+        more_than_scheduled_delivery_datetime = (
+            scheduled_message.scheduled_timestamp + datetime.timedelta(minutes=1)
+        )
+        with mock.patch(
+            "zerver.actions.scheduled_messages.timezone_now",
+            return_value=more_than_scheduled_delivery_datetime,
+        ):
+            scheduled_message = self.last_scheduled_message()
+            scheduled_message.realm.deactivated = True
+            scheduled_message.realm.save()
+            self.verify_deliver_scheduled_message_failure(scheduled_message, logger)
+
+    def test_sender_deactivated_failed_to_deliver_scheduled_message(self) -> None:
+        logger = mock.Mock()
+        self.create_scheduled_message()
+        scheduled_message = self.last_scheduled_message()
+
+        more_than_scheduled_delivery_datetime = (
+            scheduled_message.scheduled_timestamp + datetime.timedelta(minutes=1)
+        )
+        with mock.patch(
+            "zerver.actions.scheduled_messages.timezone_now",
+            return_value=more_than_scheduled_delivery_datetime,
+        ):
+            scheduled_message = self.last_scheduled_message()
+            change_user_is_active(scheduled_message.sender, False)
+            self.verify_deliver_scheduled_message_failure(scheduled_message, logger)
+
+    def test_sender_deactivated_failed_to_deliver_scheduled_message_unknown_exception(self) -> None:
+        logger = mock.Mock()
+        self.create_scheduled_message()
+        scheduled_message = self.last_scheduled_message()
+
+        more_than_scheduled_delivery_datetime = (
+            scheduled_message.scheduled_timestamp + datetime.timedelta(minutes=1)
+        )
+        with mock.patch(
+            "zerver.actions.scheduled_messages.timezone_now",
+            return_value=more_than_scheduled_delivery_datetime,
+        ):
+            scheduled_message = self.last_scheduled_message()
+            scheduled_message.delivery_type = ScheduledMessage.REMIND
```

Comment:
It might be better to just make this value `-1`, so that this test will still be valid if/when we re-add reminder support.

Reply:
I wasn't sure what is the correct way to do this, so left it out for now.